qme/exceptions.py

```diff
@@ -0,0 +1,177 @@
+"""
+
+Copyright (C) 2020 Vanessa Sochat.
+
+This Source Code Form is subject to the terms of the
+Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
+with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""
+
+
+class MissingEnvironmentVariable(RuntimeError):
+    """Thrown if a required environment variable is not provided.
+    """
+
+    def __init__(self, varname, *args, **kwargs):
+        super(MissingEnvironmentVariable, self).__init__(*args, **kwargs)
+        self.varname = varname
+
+    def __str__(self):
+        return "Missing environment variable '{}' is required".format(self.varname)
+
+
+class InvalidDatabaseFilename(RuntimeError):
+    """Database filename is invalid (must end in .db
+    """
+
+    def __init__(self, filename, *args, **kwargs):
+        super(InvalidDatabaseFilename, self).__init__(*args, **kwargs)
+        self.filename = filename
+
+    def __str__(self):
+        return "Invalid database file for sqlite, '{}' does not end in .db".format(
+            self.filename
+        )
+
+
+class DirectoryNotFoundError(FileNotFoundError):
+    """Thrown if a directory is not found
+    """
+
+    def __init__(self, dirname, reason, *args, **kwargs):
+        super(DirectoryNotFoundError, self).__init__(*args, **kwargs)
+        self.dirname = dirname
+        self.reason = reason
+
+    def __str__(self):
+        return "{} {}.".format(self.dirname, self.reason)
+
+
+class MultipleTasksExistError(RuntimeError):
+    """Thrown if multiple tasks exist.
+    """
+
+    def __init__(self, taskid, *args, **kwargs):
+        super(MultipleTasksExistError, self).__init__(*args, **kwargs)
+        self.taskid = taskid
+
+    def __str__(self):
+        return "More than one task found for {}.".format(self.taskid)
+
+
+class TaskNotFoundError(RuntimeError):
+    """Thrown if task does not exist.
+    """
+
+    def __init__(self, taskid, *args, **kwargs):
+        super(TaskNotFoundError, self).__init__(*args, **kwargs)
+        self.taskid = taskid
+
+    def __str__(self):
+        return "Cannot find task {}.".format(self.taskid)
+
+
+class NoTasksError(RuntimeError):
+    """Thrown if tasks are requested, but there are none
+    """
+
+    def __str__(self):
+        return "There are no tasks in the database."
+
+
+class MissingImportError(ImportError):
+    """Thrown if a library is missing.
+    """
+
+    def __init__(self, name, reason, *args, **kwargs):
+        super(MissingImportError, self).__init__(*args, **kwargs)
+        self.name = name
+        self.reason = reason
+
+    def __str__(self):
+        return "{} {}.".format(self.name, self.reason)
+
+
+class YerAnIdiotHarryError(RuntimeError):
```

Comment:
I would not have so much fun!  Exception names are information and they are API.  Should be informative and relatively "sparse", i.e. not a dedicated exception for every possible place where things could be raising.

Reply:
hehe, yeah this was more of just putting in there to give you a laugh, I'll remove it :)

Reply:
but seriously, if only we had these in real life! It would be much easier to raise a YouForgotToLocktheDoorAgain exception and have a large bird fly and poopland on the perpetrator than having to either lock it myself or remind said perpetrator that locking doors is a good idea. :)